### DISCUSSION TITLE
Unable to install Firedrake on Mac

### DISCUSSION THREAD

**Discussion by huisophiewang** [firedrakeproject/firedrake#3465] [Firedrake support] [answered]
Hi,
I'm having some difficulties installing Firedrake. It looks like petsc configuration failed. I'm installing it on a Macbook M3 Pro with macOS 14. I've attached the log files here. Any help would be appreciated! Thank you!
[firedrake-install.log](https://github.com/firedrakeproject/firedrake/files/14672250/firedrake-install.log)
[configure.log](https://github.com/firedrakeproject/firedrake/files/14672263/configure.log)

**Comment by connorjward** [comment]
You are failing to build chaco. It isn't essential for Firedrake to function so can you try running
```
export PETSC_CONFIGURE_OPTIONS=--download-chaco=no
```
before calling `firedrake-install`? Hopefully that will skip the build of chaco.

@JDBetteridge any insights as to why there are errors like:
```
submain/balance.c:254:18: warning: passing arguments to 'make_term_props' without a prototype is deprecated in all versions of C and is not supported in C2x [-Wdeprecated-non-prototype]
                make_term_props(graph, subnvtxs, loc2glob, assignment,
```

**Reply by JDBetteridge** [reply]
Chaco is (currently) a hard dependency, but the issue @Ig-dolci has linked contains the solution.

**Comment by Ig-dolci** [comment]
That is a petsc [issue](https://gitlab.com/petsc/petsc/-/issues/1557). So, perhaps trying to install without Chaco can be a plausible option.

**Comment by huisophiewang** [comment]
I added "--download-chaco-commit=HEAD" to chaco configuration (assuming that's the suggestion of the post referred to by @Ig-dolci), it still has the same error. 

I can potentially switch to other operations systems such as Linux or Windows. What would be the recommended operation system to run firedrake? 
[configure.log](https://github.com/firedrakeproject/firedrake/files/14706170/configure.log)

**Reply by connorjward** [reply]
If all goes to plan we should have updated our PETSc fork by the end of the day, which should resolve these issues ideally. Firedrake also works well on Linux or using the Windows Subsystem for Linux (WSL) on Windows.

**Reply by huisophiewang** [reply]
Thank you! Please let me know when PETSc is updated and what I need to do to.

**Reply by connorjward** [reply]
Sorry for not notifying you sooner. We did update PETSc today. If you delete your previous virtual environment and run `firedrake-install` again hopefully things will work for you this time.

**Reply by huisophiewang** [reply]
It still has the same Chaco error. Is there some way for me to check and make sure that I am installing the latest fixed version?

**Reply by connorjward** [reply]
Sorry it seems like PETSc have not yet fixed the Chaco build problems (https://gitlab.com/petsc/petsc/-/issues/1557). We will try to fix this ASAP next week.

**Comment by javjrg** [comment]
Hi, I have the same error. Are there some solution for this problem?

Thanks a lot for help

**Reply by connorjward** [reply]
Sorry it seems like PETSc have not yet fixed the Chaco build problems (https://gitlab.com/petsc/petsc/-/issues/1557). We will try to fix this ASAP next week.

**Comment by huisophiewang** [comment]
Is there any update about the Chaco problem?

**Reply by connorjward** [reply, marked as answer]
There appears to have been no progress made by PETSc on this. We may however have a workaround. A very helpful user (@lrtfm) on Slack said:

If you prefer not to downgrade your Xcode, here are some alternative ways to overcome the issues.
Edit the script firedrake-install  at line786: https://github.com/firedrakeproject/firedrake/blob/1bbd9dfa3b9a7dc7e501cc094b93067e89c6448c/scripts/firedrake-install#L786

```
            petsc_options.add("--CFLAGS=-Wno-implicit-function-declaration")
```
to
```
            petsc_options.add("--CFLAGS=-Wno-implicit-function-declaration -Wno-int-conversion -Wno-deprecated-non-prototype -Wno-implicit-int")
```
to overcome the issue for now.
Here, `-Wno-deprecated-non-prototype -Wno-implicit-int` for chaco, and `-Wno-int-conversion` for openblas.

2. If you encounter an error with Cython when compiling Cython code in Firedrake, you may need to apply the following patch to Firedrake.
```
diff --git a/setup.py b/setup.py
index fae92d9a7..cb21b3bff 100644
--- a/setup.py
+++ b/setup.py
@@ -68,11 +68,13 @@ link_args = ["-L%s/lib" % d for d in dirs] + ["-Wl,-rpath,%s/lib" % d for d in d
 libspatialindex_so = Path(rtree.core.rt._name).absolute()
 link_args += [str(libspatialindex_so)]
 link_args += ["-Wl,-rpath,%s" % libspatialindex_so.parent]
+extra_compile_args = ["-Wno-incompatible-function-pointer-types"]
 
 extensions = [Extension("firedrake.cython.{}".format(ext),
                         sources=[os.path.join("firedrake", "cython", "{}.pyx".format(ext))],
                         include_dirs=include_dirs,
                         libraries=libs,
+                        extra_compile_args=extra_compile_args,
                         extra_link_args=link_args,
                         cython_compile_time_env=cython_compile_time_env) for (ext, libs) in cythonfiles]
 if 'CC' not in env:
```

**Reply by lrtfm** [reply]
There appears to be a segmentation fault issue with MUMPS in the current PETSc version on MacOS (arm64) when running __*in parallel*__, see [here](https://lists.mcs.anl.gov/pipermail/petsc-users/2024-March/050482.html).

In addition to the previously mentioned patches, if you intend to use MUMPS in parallel, you may want to consider using the [latest version of PETSc](https://github.com/firedrakeproject/petsc/tree/Firedrake_20240312.0) like so:
```
python3 firedrake-install --package-branch petsc Firedrake_20240312.0
```

**Reply by javjrg** [reply]
Thanks a lot for your support and time, I followed the instruction but when I tried to update using

```bash
firedrake-update
```

I have the following error:

```bash
Running /Users/javjrg/firedrake/bin/firedrake-update
BLAS config is {'BLAS': '/Users/javjrg/firedrake/src/petsc/default', 'OPENBLAS': '/Users/javjrg/firedrake/src/petsc/default', 'LDFLAGS': '-L/opt/homebrew/opt/gcc/lib/gcc/13'}
Configuration saved to /Users/javjrg/firedrake/.configuration.json
Updating the git repository for firedrake
Creating firedrake-update script.
Running ../bin/firedrake-update --no-update-script
Installing command line tools...
Checking brew doctor...
Installing required packages via homebrew...
Required package 'gcc' is already installed via Homebrew.
Required package 'autoconf' is already installed via Homebrew.
Required package 'make' is already installed via Homebrew.
Required package 'automake' is already installed via Homebrew.
Required package 'cmake' is already installed via Homebrew.
Required package 'libtool' is already installed via Homebrew.
Required package 'boost' is already installed via Homebrew.
BLAS config is {'BLAS': '/Users/javjrg/firedrake/src/petsc/default', 'OPENBLAS': '/Users/javjrg/firedrake/src/petsc/default', 'LDFLAGS': '-L/opt/homebrew/opt/gcc/lib/gcc/13'}
Configuration saved to /Users/javjrg/firedrake/.configuration.json
Finding dependencies of PyOP2

Finding dependencies of firedrake

Updating the git repository for petsc
Traceback (most recent call last):
  File "/Users/javjrg/firedrake/src/../bin/firedrake-update", line 1964, in <module>
    update_petsc = git_update("petsc", deps["petsc"])
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/javjrg/firedrake/src/../bin/firedrake-update", line 982, in git_update
    check_call(["git", "pull", "--recurse-submodules"])
  File "/Users/javjrg/firedrake/src/../bin/firedrake-update", line 674, in check_call
    log.debug(subprocess.check_output(arguments, stderr=subprocess.STDOUT, env=os.environ).decode())
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.11/3.11.9/Frameworks/Python.framework/Versions/3.11/lib/python3.11/subprocess.py", line 466, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.11/3.11.9/Frameworks/Python.framework/Versions/3.11/lib/python3.11/subprocess.py", line 571, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['git', 'pull', '--recurse-submodules']' returned non-zero exit status 1.


Update log saved in /Users/javjrg/firedrake/firedrake-update.log
```


Can someone help me with that?

[firedrake-update.log](https://github.com/firedrakeproject/firedrake/files/14934812/firedrake-update.log)

**Reply by connorjward** [reply]
This says that you cannot run the command `git pull` inside of `firedrake/src/petsc`. Do you have unstashed changes? You can `cd $VIRTUAL_ENV/src/petsc` then `git status` to see if that's the case. You should be able to run `git pull` from that directory without error.

**Reply by javjrg** [reply]
Thanks for your answer. It looks like you've resolved the issue with the branch. To update, I used:

```bash
git checkout firedrake
```

Then, I was able to update without any problems. Finally, I ran the tests with:

```bash
cd $VIRTUAL_ENV/src/firedrake
pytest tests/regression/ -k "poisson_strong or stokes_mini or dg_advection"
```

And I received the following output:

```bash
================================================================================================ test session starts =================================================================================================
platform darwin -- Python 3.11.9, pytest-8.1.1, pluggy-1.4.0
rootdir: /Users/javjrg/firedrake/src/firedrake
configfile: setup.cfg
plugins: anyio-4.3.0, mpi-0.1, xdist-3.5.0, nbval-0.11.0
collected 3517 items / 3490 deselected / 1 skipped / 27 selected                                                                                                                                                     

tests/regression/test_dg_advection.py ....                                                                                                                                                                     [ 14%]
tests/regression/test_poisson_strong_bcs.py .................                                                                                                                                                  [ 77%]
tests/regression/test_poisson_strong_bcs_nitsche.py ....                                                                                                                                                       [ 92%]
tests/regression/test_stokes_mini.py ..                                                                                                                                                                        [100%]

================================================================================================== warnings summary ==================================================================================================
../fiat/FIAT/__init__.py:5
  /Users/javjrg/firedrake/src/fiat/FIAT/__init__.py:5: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
    import pkg_resources

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
======================================================================== 27 passed, 1 skipped, 3490 deselected, 1 warning in 96.82s (0:01:36) ========================================================================
WARNING! There are options you set that were not used!
WARNING! could be spelling mistake, etc!
There is one unused database option. It is:
Option left: name:-k value: poisson_strong or stokes_mini or dg_advection source: command line
```

It seems like everything is now in order. Thanks again for your time.

**Comment by Ig-dolci** [comment]
> There appears to be a segmentation fault issue with MUMPS in the current PETSc version on MacOS (arm64) when running **_in parallel_**, see [here](https://lists.mcs.anl.gov/pipermail/petsc-users/2024-March/050482.html).
> 
> In addition to the previously mentioned patches, if you intend to use MUMPS in parallel, you may want to consider using the [latest version of PETSc](https://github.com/firedrakeproject/petsc/tree/Firedrake_20240312.0) like so:
> 
> ```
> python3 firedrake-install --package-branch petsc Firedrake_20240312.0
> ```

I also opened an [issue](https://gitlab.com/petsc/petsc/-/issues/1569) in PETSc for the errors with MUMPS.
